binary_tree.c

```diff
@@ -0,0 +1,410 @@
+    struct Node* parent = tree_search_parent(tree, self);
+
+    if(self->l_node == NULL && self->r_node == NULL) {
+        free((void*) self);
+    }
+    else if(parent->l_node && parent->r_node == NULL) {
+        if(tree->root == self) {
+            tree->root = self->l_node;
+        }
+        else {
+            parent->l_node = self->l_node;
+        }
+    }
+    else if(parent->r_node && parent->l_node == NULL) {
+        if(tree->root == self) {
+            tree->root = self->r_node;
+        }
+        else {
+            parent->r_node = self->r_node;
+        }
+    }
+    else {
+        if (self->l_node->r_node) {
+            // find last right vertice(where node->r_node==NULL)
+            struct Node* last_right_vertice = tree_search_last_right_vertice(self->l_node);
+
+            if(last_right_vertice->l_node) {
+                self->l_node->r_node = last_right_vertice->l_node;
+
+                if(self->r_node) {
+                    last_right_vertice->r_node = self->r_node;
+                }
+                else {
+                    last_right_vertice->l_node = NULL;
+                }
+            }
+
+            last_right_vertice->l_node = self->l_node;
+            parent->l_node = last_right_vertice;
+
+            if(tree->root==self) {
+                tree->root = last_right_vertice;
+            }
+        }
+        else {
+            if(tree->root==self && tree->root->r_node) {
+                self->l_node->r_node = self->r_node;
+                tree->root = tree->root->l_node;
+            }
+            else if(parent->r_node)
+            {
+                self->l_node->r_node = self->r_node;
+                parent->l_node = self->l_node;
+            }
+        }
+    }
+    free((void*) self);
+}
+
+struct Node* tree_search_parent(Tree* tree, struct Node* self) {
+    struct Node* parent = NULL;
+
+    if(tree->root == self) {
+        parent = self;
+        return parent;
+    }
+    else if(tree->root->l_node && tree->root->l_node == self) {
+        parent = tree->root;
+        return parent;
+    }
+    else if(tree->root->r_node && tree->root->r_node == self) {
+        parent = tree->root;
+        return parent;
+    }
+    else {
+        if(tree->root->l_node) {
+            Tree left_tree = tree_create();
+            left_tree.root = tree->root->l_node;
+            parent = tree_search_parent(&left_tree, self);
+
+            if((parent) && parent->l_node == self) {
+                return parent;
+            }
+        }
+
+        if(tree->root->r_node) {
+            Tree right_tree = tree_create();
+            right_tree.root = tree->root->r_node;
+            parent = tree_search_parent(&right_tree, self);
+
+            if((parent) && parent->r_node == self) {
+                return parent;
+            }
+        }
+
+        return parent;
+    }
+}
+
+struct Node* tree_search_last_right_vertice(struct Node* left_child) {
```

Comment:
I guess it's not a public interface of a tree. So if it's meant to be used only inside binary_tree.c we usually do not put such functions to headers, and in source files mark them static